### PR TITLE
Mediapipe models subconfig

### PR DESCRIPTION
**Pull request by michalkulakowski**

### CODE REVIEW

File: src/schema.cpp [hunk 1]
```diff
@@ -416,6 +419,95 @@ const char* MODELS_MAPPING_SCHEMA = R"(
     "additionalProperties": false
 })";
 
+const char* MEDIAPIPE_CONFIG_SCHEMA = R"({
+	"definitions": {
```
Comment (atobiszei) [inline src/schema.cpp:423]: Is there no way to reuse already defined schema for models? Otherwise we have to remember to change schema in two places.

File: src/mediapipe_internal/mediapipegraphconfig.hpp [hunk 2]
```diff
@@ -53,20 +53,45 @@ class MediapipeGraphConfig {
...
 
+    /**
+     * @brief Json config path
+     */
+    std::string modelsConfigPath;
+
+    void setPath(std::string& path, const std::string& givenPath) {
+        path = givenPath;
+
+        if (!FileSystem::isLocalFilesystem(givenPath)) {
```
Comment (atobiszei) [inline src/mediapipe_internal/mediapipegraphconfig.hpp:64]: Check this in ovms repo:
grep -r -R -A8 -B3 isLocalFilesystem

we already have similar function bodies in mediapipegraphconfig.hpp, customloaderconfig.hpp, modelmanager.cpp, modelconfig.hpp. Maybe we can have utility function? @rasapala

File: src/modelmanager.cpp [hunk 3]
```diff
@@ -366,9 +366,9 @@ static Status processCustomNodeConfig(const rapidjson::Value& nodeConfig, Custom
     return StatusCode::OK;
 }
 
-static Status processMediapipeConfig(rapidjson::Document& configJson, const rapidjson::Value& pipelineConfig, std::set<std::string>& mediapipesInConfigFile, MediapipeFactory& factory, ModelManager& manager) {
+Status ModelManager::processMediapipeConfig(rapidjson::Document& configJson, const rapidjson::Value& pipelineConfig, std::set<std::string>& mediapipesInConfigFile, MediapipeFactory& factory, std::vector<ModelConfig>& gatedModelConfigs) {
```
Comment (atobiszei) [inline src/modelmanager.cpp:369]: Why we need to make this method instead of function? Idea here is that eventually we should separate configuration parsing logic from ModelManager so at least this function was ready.

File: src/test/mediapipeflow_test.cpp [hunk 4]
```diff
@@ -90,6 +90,35 @@ class MediapipeFlowDummyTest : public MediapipeFlowTest {
...
+    ::KFSResponse response;
+
+    const std::string modelName = GetParam();
+    request.Clear();
+    response.Clear();
+    inputs_info_t inputsMeta{{"in", {DUMMY_MODEL_SHAPE, precision}}};
+    preparePredictRequest(request, inputsMeta);
+    request.mutable_model_name()->assign(modelName);
+    ASSERT_EQ(impl.ModelInfer(nullptr, &request, &response).error_code(), grpc::StatusCode::OK);
+    // TODO validate output
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:113]: Use checkDummyResponse?

File: ci/cppclean.sh [hunk 5]
```diff
@@ -43,7 +43,7 @@ if [ ${NO_WARNINGS_NOTUSED} -gt 7 ]; then
     exit 1;
 fi
 
-if [ ${NO_WARNINGS} -gt  161 ]; then
+if [ ${NO_WARNINGS} -gt  162 ]; then
```
Comment (atobiszei) [inline ci/cppclean.sh:46]: Why you need to increase threshold?

**Review by rasapala — APPROVED** [review]
LGTM

File: src/mediapipe_internal/mediapipegraphconfig.hpp [hunk 6]
```diff
@@ -108,20 +116,25 @@ class MediapipeGraphConfig {
...
-        } else if (graphPath.at(0) == '/') {
-            // Full path case
-            this->graphPath = graphPath;
-        } else {
-            // Relative path case
-            if (this->rootDirectoryPath.empty())
-                throw std::logic_error("Using graph relative path without setting configuration directory path.");
-            this->graphPath = this->rootDirectoryPath + graphPath;
-        }
+        FileSystem::setPath(this->graphPath, graphPath, this->rootDirectoryPath);
```
Comment (atobiszei) [inline src/mediapipe_internal/mediapipegraphconfig.hpp:119]: Move implementation to cpp file and then we don't need to include filesystem here.

File: src/schema.cpp [hunk 7]
```diff
@@ -416,6 +421,23 @@ const char* MODELS_MAPPING_SCHEMA = R"(
     "additionalProperties": false
 })";
 
+const std::string MEDIAPIPE_CONFIG_SCHEMA = R"({
```
Comment (atobiszei) [inline src/schema.cpp:424]: Rename?
```suggestion
const std::string MEDIAPIPE_SUBCONFIG_SCHEMA = R"({
```

File: src/modelmanager.cpp [hunk 8]
```diff
@@ -750,9 +759,50 @@ Status ModelManager::loadModelsConfig(rapidjson::Document& configJson, std::vect
...
+            SPDLOG_LOGGER_ERROR(modelmanager_logger, "Mediapipe configuration file is not in valid configuration format");
+            return StatusCode::JSON_INVALID;
+        }
+        const auto mediapipeItr = configJson.FindMember("model_config_list");
+
+        if (mediapipeItr == configJson.MemberEnd() || !mediapipeItr->value.IsArray()) {
+            SPDLOG_LOGGER_ERROR(modelmanager_logger, "Configuration file doesn't have models property.");
+            return StatusCode::JSON_INVALID;
+        }
+        auto status = loadModels(mediapipeItr, gatedModelConfigs, modelsInConfigFile, modelsWithInvalidConfig, newModelConfigs);
```
Comment (atobiszei) [inline src/modelmanager.cpp:801]: We could report here for which mediapipe loading subconfig didn't succeed

Comment (atobiszei) [reply src/modelmanager.cpp:801]: Do we report if subconfig contains model name that is already loaded?

File: src/BUILD [hunk 9]
```diff
@@ -366,6 +366,7 @@ cc_library(
                 "mediapipe_internal/mediapipefactory.cpp",
                 "mediapipe_internal/mediapipefactory.hpp",
                 "mediapipe_internal/mediapipegraphconfig.hpp",
+                "mediapipe_internal/mediapipegraphconfig.cpp",
```
Comment (atobiszei) [inline src/BUILD:369]: Add new config files to data files in ovms_test

File: src/test/mediapipeflow_test.cpp [hunk 10]
```diff
@@ -95,6 +95,66 @@ class MediapipeFlowDummyTest : public MediapipeFlowTest {
...
 };
 
+class MediapipeFlowDummySeparateConfigTest : public MediapipeFlowTest {
+public:
+    void SetUp() {
+        SetUpServer("/ovms/src/test/mediapipe/config_mediapipe_dummy_adapter_full_separate_config.json");
+    }
+};
+
+TEST_P(MediapipeFlowDummySeparateConfigTest, Infer) {
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:105]: Why this test needs to be parametrized?

Comment (atobiszei) [reply src/test/mediapipeflow_test.cpp:105]: Other tests were parametrized to verify support for 1 or more inputs (dummy vs add). I don't see need here?

File: src/test/mediapipeflow_test.cpp [hunk 11]
```diff
@@ -95,6 +95,66 @@ class MediapipeFlowDummyTest : public MediapipeFlowTest {
...
+}
+
+class MediapipeFlowDummyDummyInSubconfigAndConfigTest : public MediapipeFlowTest {
+public:
+    void SetUp() {
+        SetUpServer("/ovms/src/test/mediapipe/config_mediapipe_dummy_adapter_full_dummy_in_both_config_and_subconfig.json");
+    }
+};
+
+TEST_P(MediapipeFlowDummyDummyInSubconfigAndConfigTest, Infer) {
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:135]: What happens in this scenario? We could have in subconfig model dummy with shape (1,12), and in main config shape: (1,10). Which model will be loaded and if there will be warning/information in logs that would ease debug?

File: src/test/mediapipe/config_mediapipe_dummy_adapter_full_dummy_in_both_config_and_subconfig.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    ],
+    "mediapipe_config_list": [
+    {
+        "name":"mediaDummyADAPTFULL",
+        "graph_path":"/ovms/src/test/mediapipe/graphdummyadapterfull.pbtxt",
+        "subconfig":"/ovms/src/test/mediapipe/config_standard_dummy.json"
+    },
+    {
+        "name":"mediaDummy",
```
Comment (atobiszei) [inline src/test/mediapipe/config_mediapipe_dummy_adapter_full_dummy_in_both_config_and_subconfig.json:17]: I assume we could change one of those graphs to use base_path without passing full graph_path/subconfig path when base_path support is added?

File: src/test/mediapipeflow_test.cpp [hunk 13]
```diff
@@ -614,10 +718,65 @@ TEST_F(MediapipeConfigChanges, AddImroperGraphThenFixWithReloadThenBreakAgain) {
     modelManager.loadConfig(configFilePath);
     definition = factory.findDefinitionByName(mgdName);
     ASSERT_NE(nullptr, definition);
-    ASSERT_EQ(definition->getStatus().getStateCode(), PipelineDefinitionStateCode::LOADING_PRECONDITION_FAILED);
+    ASSERT_EQ(definition->getStatus().getStateCode(), PipelineDefinitionStateCode::RETIRED);
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:721]: Why this change is required?

Comment (atobiszei) [reply src/test/mediapipeflow_test.cpp:721]: Here the test should show that changing back to improper config will cause the reload and the mediapipe will be in failed load state.

File: src/test/mediapipeflow_test.cpp [hunk 14]
```diff
@@ -614,10 +718,65 @@ TEST_F(MediapipeConfigChanges, AddImroperGraphThenFixWithReloadThenBreakAgain) {
...
+    const std::string modelPathToReplace{"XYZ"};
+    configFileContent.replace(configFileContent.find(modelPathToReplace), modelPathToReplace.size(), graphFilePath);
+    createConfigFileWithContent(configFileContent, configFilePath);
+    createConfigFileWithContent(pbtxtContent, graphFilePath);
+    ConstructorEnabledModelManager modelManager;
+    modelManager.loadConfig(configFilePath);
+    const MediapipeFactory& factory = modelManager.getMediapipeFactory();
+    auto model = modelManager.findModelByName("dummy");
+    ASSERT_NE(nullptr, model->getDefaultModelInstance());
+    ASSERT_EQ(model->getDefaultModelInstance()->getStatus().getErrorCode(), ModelVersionStatusErrorCode::OK);
```
Comment (atobiszei) [inline src/test/mediapipeflow_test.cpp:738]: Here test for getStateCode(). getErrorCode is useful only for checking the reason of failed load.

File: src/test/mediapipeflow_test.cpp (hunk 11 above)
Comment (atobiszei) [reply src/test/mediapipeflow_test.cpp:135]: I think that the subconfig model will be loaded. For now we can leave it as known issue since this is preview but ideally we should report issue that you cannot have the same model configured in main json & subconfig.
For now add test documenting the behaviour whichever it is (expecting shape from subconfig or main config).

**Review by atobiszei — APPROVED** [review]
